In [1]:
import pandas as pd
import numpy as np

In [2]:
df= pd.read_csv("D:/Datasets/mercedes.csv",usecols=['X1','X2','X3','X4','X5','X6'])

In [17]:
df.shape

(4209, 6)

In [3]:
df.head()

X1  X2 X3 X4 X5 X6
0  v  at  a  d  u  j
1  t  av  e  d  y  l
2  w   n  c  d  x  j
3  t   n  f  d  x  l
4  v   n  f  d  h  d

In [4]:
## Lets have a look at how many unique labels each variable has.

for col in df.columns:
    print(f'{col} : {len(df[col].unique())} labels')

X1 : 27 labels
X2 : 44 labels
X3 : 7 labels
X4 : 4 labels
X5 : 29 labels
X6 : 12 labels


In [6]:
## Let's examine how many columns we will obtain after one hot encoding these variables.
pd.get_dummies(df,drop_first=True)

X1_aa  X1_ab  X1_b  X1_c  X1_d  X1_e  X1_f  X1_g  X1_h  X1_i  ...  X6_c  \
0         0      0     0     0     0     0     0     0     0     0  ...     0   
1         0      0     0     0     0     0     0     0     0     0  ...     0   
2         0      0     0     0     0     0     0     0     0     0  ...     0   
3         0      0     0     0     0     0     0     0     0     0  ...     0   
4         0      0     0     0     0     0     0     0     0     0  ...     0   
...     ...    ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
4204      0      0     0     0     0     0     0     0     0     0  ...     0   
4205      0      0     0     0     0     0     0     0     0     0  ...     0   
4206      0      0     0     0     0     0     0     0     0     0  ...     0   
4207      0      0     0     0     0     0     0     0     0     0  ...     0   
4208      0      0     0     0     0     0     0     0     0     0  ...     0   

      X6_d  X6_e  X6_f  X6_g  X6_h  X6_i  X6_j  X6_k  X6_l  
0        0     0     0     0     0     0     1     0     0  
1        0     0     0     0     0     0     0     0     1  
2        0     0     0     0     0     0     1     0     0  
3        0     0     0     0     0     0     0     0     1  
4        1     0     0     0     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...   ...   ...   ...  
4204     1     0     0     0     0     0     0     0     0  
4205     0     0     0     0     1     0     0     0     0  
4206     0     0     0     1     0     0     0     0     0  
4207     0     0     0     0     0     0     0     0     1  
4208     0     0     0     1     0     0     0     0     0  

[4209 rows x 117 columns]

We can see that from just 6 initital categorical variables, we end up with 117 new variables.

What can we do??

We can find out 10 most frequent categories in each feature and use that, and skip the other categories in each feature.<br>
For all the remaining categories it will be zero.

In [22]:
## Let's find the top 10 most frequent categories for the variable x2
df.X2.value_counts().sort_values(ascending=False).head(10)

as    1659
ae     496
ai     415
m      367
ak     265
r      153
n      137
s       94
f       87
e       81
Name: X2, dtype: int64

In [28]:
## Let's make a list with the most frequent categories of the variable.

top_10 = [x for x in df.X2.value_counts().sort_values(ascending=False).head(10).index]

In [29]:
top_10

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

In [32]:
## Now we make 10 binary variables.

for label in top_10:
    df[label] = np.where(df['X2']==label,1,0)
    
df[['X2']+top_10].head(40)

X2  as  ae  ai  m  ak  r  n  s  f  e
0   at   0   0   0  0   0  0  0  0  0  0
1   av   0   0   0  0   0  0  0  0  0  0
2    n   0   0   0  0   0  0  1  0  0  0
3    n   0   0   0  0   0  0  1  0  0  0
4    n   0   0   0  0   0  0  1  0  0  0
5    e   0   0   0  0   0  0  0  0  0  1
6    e   0   0   0  0   0  0  0  0  0  1
7   as   1   0   0  0   0  0  0  0  0  0
8   as   1   0   0  0   0  0  0  0  0  0
9   aq   0   0   0  0   0  0  0  0  0  0
10   r   0   0   0  0   0  1  0  0  0  0
11   e   0   0   0  0   0  0  0  0  0  1
12  ai   0   0   1  0   0  0  0  0  0  0
13   e   0   0   0  0   0  0  0  0  0  1
14  as   1   0   0  0   0  0  0  0  0  0
15  ak   0   0   0  0   1  0  0  0  0  0
16   e   0   0   0  0   0  0  0  0  0  1
17  as   1   0   0  0   0  0  0  0  0  0
18   m   0   0   0  1   0  0  0  0  0  0
19  ak   0   0   0  0   1  0  0  0  0  0
20   m   0   0   0  1   0  0  0  0  0  0
21  as   1   0   0  0   0  0  0  0  0  0
22   a   0   0   0  0   0  0  0  0  0  0
23   m   0   0   0  1   0  0  0  0  0  0
24  ak   0   0   0  0   1  0  0  0  0  0
25   m   0   0   0  1   0  0  0  0  0  0
26  as   1   0   0  0   0  0  0  0  0  0
27  as   1   0   0  0   0  0  0  0  0  0
28  as   1   0   0  0   0  0  0  0  0  0
29   m   0   0   0  1   0  0  0  0  0  0
30  as   1   0   0  0   0  0  0  0  0  0
31  as   1   0   0  0   0  0  0  0  0  0
32   m   0   0   0  1   0  0  0  0  0  0
33  as   1   0   0  0   0  0  0  0  0  0
34  as   1   0   0  0   0  0  0  0  0  0
35  ai   0   0   1  0   0  0  0  0  0  0
36  as   1   0   0  0   0  0  0  0  0  0
37   k   0   0   0  0   0  0  0  0  0  0
38  as   1   0   0  0   0  0  0  0  0  0
39  as   1   0   0  0   0  0  0  0  0  0

In [40]:
### get whole set of dummy variables for all categorical variables.

def one_hot_top_x(df,variable,top_x_labels):
    ## Function to create dummy variables for the most frequent labels.
    ## We can vary the number of most frequent labels that we encode.
    
    for label in top_x_labels:
        df[variable+'_'+label] = np.where(df[variable]==label,1,0)
        
# read the data again.
df = pd.read_csv("D:/Datasets/mercedes.csv",usecols=['X1','X2','X3','X4','X5','X6'])

# find the top_10 labels manually and then pass it to the above function.
top_10 = [x for x in df['X2'].value_counts().sort_values(ascending=False).head(10).index]

# encode X2 into the 10 most frequent categories.
one_hot_top_x(df,'X2',top_10)

In [57]:
df

X1  X2 X3 X4  X5 X6  X2_as  X2_ae  X2_ai  X2_m  X2_ak  X2_r  X2_n  X2_s  \
0     v  at  a  d   u  j      0      0      0     0      0     0     0     0   
1     t  av  e  d   y  l      0      0      0     0      0     0     0     0   
2     w   n  c  d   x  j      0      0      0     0      0     0     1     0   
3     t   n  f  d   x  l      0      0      0     0      0     0     1     0   
4     v   n  f  d   h  d      0      0      0     0      0     0     1     0   
...  ..  .. .. ..  .. ..    ...    ...    ...   ...    ...   ...   ...   ...   
4204  s  as  c  d  aa  d      1      0      0     0      0     0     0     0   
4205  o   t  d  d  aa  h      0      0      0     0      0     0     0     0   
4206  v   r  a  d  aa  g      0      0      0     0      0     1     0     0   
4207  r   e  f  d  aa  l      0      0      0     0      0     0     0     0   
4208  r  ae  c  d  aa  g      0      1      0     0      0     0     0     0   

      X2_f  X2_e  
0        0     0  
1        0     0  
2        0     0  
3        0     0  
4        0     0  
...    ...   ...  
4204     0     0  
4205     0     0  
4206     0     0  
4207     0     1  
4208     0     0  

[4209 rows x 16 columns]

In [58]:
top_10_X1 = [x for x in df['X1'].value_counts().sort_values(ascending=False).head().index]

In [59]:
one_hot_top_x(df,'X1',top_10_X1)

In [60]:
df

X1  X2 X3 X4  X5 X6  X2_as  X2_ae  X2_ai  X2_m  ...  X2_r  X2_n  X2_s  \
0     v  at  a  d   u  j      0      0      0     0  ...     0     0     0   
1     t  av  e  d   y  l      0      0      0     0  ...     0     0     0   
2     w   n  c  d   x  j      0      0      0     0  ...     0     1     0   
3     t   n  f  d   x  l      0      0      0     0  ...     0     1     0   
4     v   n  f  d   h  d      0      0      0     0  ...     0     1     0   
...  ..  .. .. ..  .. ..    ...    ...    ...   ...  ...   ...   ...   ...   
4204  s  as  c  d  aa  d      1      0      0     0  ...     0     0     0   
4205  o   t  d  d  aa  h      0      0      0     0  ...     0     0     0   
4206  v   r  a  d  aa  g      0      0      0     0  ...     1     0     0   
4207  r   e  f  d  aa  l      0      0      0     0  ...     0     0     0   
4208  r  ae  c  d  aa  g      0      1      0     0  ...     0     0     0   

      X2_f  X2_e  X1_aa  X1_s  X1_b  X1_l  X1_v  
0        0     0      0     0     0     0     1  
1        0     0      0     0     0     0     0  
2        0     0      0     0     0     0     0  
3        0     0      0     0     0     0     0  
4        0     0      0     0     0     0     1  
...    ...   ...    ...   ...   ...   ...   ...  
4204     0     0      0     1     0     0     0  
4205     0     0      0     0     0     0     0  
4206     0     0      0     0     0     0     1  
4207     0     1      0     0     0     0     0  
4208     0     0      0     0     0     0     0  

[4209 rows x 21 columns]

# Advantages.

1. Straightforward to implement
2. Does not reuire hrs of variable exploration
3. Does not expand maassively the feature space(number of columns in the dataset)

# Disadvantages.

1. Does not add any information that make the variable more predictive.
2. Does not keep the information of the ignored labels.

Because it is not unusual that categorical variables have a few domaination categories and the remaining labels add mostly
noise, this is a qiote simple and straighforward approach that may be useful on many occassions. <br>
<br>
It is worth nothing that the top 10 variables is totally arbitary number. You could also choose top 5 or top 20.